Improve error checking

jcfr opened the pull request:
### Summary

* Generalize handling of contractions like "'d", "'s" or "'th"
* Perform camel case splitting after removing the prefix
  * Since a given prefix may not have been added to the dictionary, adds an explicit check to verify that the remaining of the word is valid after stripping the prefix.
* Output readability improvements
* Reduce complexity introducing functions like `spell_check_comment` and `checkWords`
* Miscellaneous optimization

_Note that additional test cases have also been introduced._

### Output

**After:**

```
Additional dictionary: /home/jcfr/Projects/SimpleITKSpellChecking/additional_dictionary.txt
Additional dictionary: tests/dict.txt
Prefixes: ['sitk', 'itk', 'vtk', 'myprefix']
Suffixes: ['.h']

Checking tests/example.h
spell_check_file: tests/example.h, text/x-c++
Line 3:  Mary had a little lamb
Line 4: 
Line 5:  Prefix test words:
Line 6:  sitkBanana
Error: sitkBanana
Trying without 'sitk' prefix: sitkBanana -> Banana
Line 7:  vtkApple
Error: vtkApple
Trying without 'vtk' prefix: vtkApple -> Apple
Line 8:  itkPineapple
Error: itkPineapple
Trying without 'itk' prefix: itkPineapple -> Pineapple
Line 9: 
Line 10:  Camel case test word:
Line 11:  CamelCaseTestWord
[...]
```

**Before:**

```
Prefixes: ['sitk', 'itk', 'vtk']
Suffixes: ['.h']

Checking tests/example.h
spell_check_file: tests/example.h , text/x-c++
Comment:   Mary had a little lamb
<class 'comment_parser.parsers.common.Comment'>
Comment:  
<class 'comment_parser.parsers.common.Comment'>
Comment:   Prefix test words:
<class 'comment_parser.parsers.common.Comment'>
Comment:   sitkBanana
<class 'comment_parser.parsers.common.Comment'>
Error: sitkBanana
Trying without prefix:  sitkBanana Banana
Trying splitting camel case word
Comment:   vtkApple
<class 'comment_parser.parsers.common.Comment'>
Error: vtkApple
Trying without prefix:  vtkApple Apple
Trying splitting camel case word
Comment:   itkPineapple
<class 'comment_parser.parsers.common.Comment'>
Error: itkPineapple
Trying without prefix:  itkPineapple Pineapple
Trying splitting camel case word
Comment:  
<class 'comment_parser.parsers.common.Comment'>
Comment:   Camel case test word:
<class 'comment_parser.parsers.common.Comment'>
Comment:   CamelCaseTestWord
<class 'comment_parser.parsers.common.Comment'>
Error: CamelCaseTestWord
```

jcfr commented:
I will revisit how contraction like `'s`, `'d` are handled.

jhlegarreta commented on codespell.py:
```diff
@@ -72,6 +73,14 @@ def load_text_file(filename):
 
 
+def checkWords(spell_checker, words):
+    """Check each words and report False if at least one has error."""
```
words : word

jhlegarreta reviewed (commented):
Thanks for the effort @jcfr. Have not tried.

blowekamp commented:
@jcfr Looks like my black formatting conflicted with this PR. And it was bigger than I though it was. 

I rebased and applied black formatting to the PR here: https://github.com/blowekamp/SimpleITKSpellChecking/tree/jcfr-improve-error-checking

jcfr commented:
@jhlegarreta @blowekamp @dave3d  This is ready for review and integration. Thanks for your time.

jhlegarreta commented on codespell.py:
```diff
@@ -80,14 +81,108 @@ def load_text_file(filename):
     return output
 
 
+def checkWords(spell_checker, words):
```
Prefer using Python convection (`lower_case_with_underscores`)?

jcfr replied:
For future reference:
* #37

jhlegarreta commented on codespell.py:
```diff
@@ -80,14 +81,108 @@ def load_text_file(filename):
+        valid = False
+
+        # Check for contractions
+        for contraction in ["'d", "'s", "'th"]:
```
IMO, it is usually cleaner to have constants such as these defined at the beginning, so that we do not find them deep into the code, may be easier to locate or re-use. Maybe a preference.

jcfr replied:
See the following pull request:
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/44

jhlegarreta reviewed (commented):
Thanks for doing this @jcfr. Have not tested, but looks good to me.

A few style-related inline comments.

One thing that I'd do is to reduce the number of commits while doing one commit per purpose:  e.g. prefer f-strings; make docstrings consistent; user message improvements; changes to the logic/refactoring. I know it is hard, and at times there is no clear boundary across them, but usually helps when reading the changes/history. If you think the current commits do already this, then it's fine, but IMO, it is hard to tell right know because commits do not have a message body either, so the subject is not enough to explain the rationale behind.

jcfr commented:
> One thing that I'd do is to reduce the number of commits while doing one commit per purpose:

To follow-up on this, here some independent pull request ready for review and integration:
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/31
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/32
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/33
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/34
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/35
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/36

Once these are integrated, I will follow with additional ones.

jhlegarreta commented:
I still miss a commit message body across some of the commits, e.g. PRs #31 and 32; at the very least it could be a copy of the subject without the prefix. Thanks for the effort.

jcfr commented:
@jhlegarreta Thanks for the feedback, I integrated most of your recommendation.

For reference, here are additional pull requests that have been derived from this one:
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/37
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/38
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/39
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/42

jcfr commented:
> commit message 

Thanks for the suggestion. When relevant, additional details were included in the body.

jcfr commented:
And here the last pull request of the series:
* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/45

This pull request can now be closed.

jhlegarreta commented:
Thanks for the effort and all this work @jcfr. I know adding a commit message body requires investing time, but IMO it is essential for reviewers and when we need to go back to read the code, it becomes instrumental to know why the change was made. Also, what is apparent to oneself may not be for others.

Dave merged all PRs now.

jcfr commented:
Agreed. I tried to find a reasonable balance.

To help in the future, for the few commits that didn't have a "body", I added details to the pull request description:

* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/31
  * Added text `Describing methods helps developer better understand the purpose of a function.`

* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/32
  * Added text `Original description of the argument didn't provide context.`

* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/37
  *  Added text `Refactor introducing a smaller function to help developer better understand and maintain the code.`

* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/42
  *  Added text
    ```
    * This pull request has the following purposes:
      * update the naming from `possessive` to the more general terminology called `contraction`
      * add support for additional contractions
      * systematically strip the known contraction before doing any error checking and accept the updated word if associated spell check succeeds after removing the contraction.
    ```

* https://github.com/SimpleITK/SimpleITKSpellChecking/pull/43
  * Added text `To avoid performing unneeded additional spell check by removing all prefixes, this pull request set the `valid` internal variable to `True` and exit the loop if the error word without a given prefix is valid.`
